compiler/src/dotty/tools/dotc/transform/SpecializeFunctions.scala

```diff
@@ -0,0 +1,231 @@
+          }
+          val specializedMethodName = specializedName(nme.apply, args, ret)
+          val specializedApply = ctx.owner.info.decls.lookup(specializedMethodName)
+
+          if (specializedApply.exists)
+            Some(orig -> (specializedParent, specializedApply.asTerm))
+          else None
+        }).flatten.toMap
+
+        val body0 = tmpl.body.foldRight(List.empty[Tree]) {
+          case (tree: DefDef, acc) if tree.name == nme.apply => {
+            val inheritedFrom =
+              tree.symbol.allOverriddenSymbols
+              .map(_.owner)
+              .map(symbolMap.get)
+              .flatten
+              .toList
+              .headOption
+
+            inheritedFrom.map { case (parent, apply) =>
+              val forwardingBody = tpd
+                .ref(apply)
+                .appliedToArgs(tree.vparamss.head.map(vparam => ref(vparam.symbol)))
+
+              val applyWithForwarding = cpy.DefDef(tree)(rhs = forwardingBody)
+
+              val specializedApplyDefDef =
+                polyDefDef(apply, trefs => vrefss => {
+                  tree.rhs
+                    .changeOwner(tree.symbol, apply)
+                    .subst(tree.vparamss.flatten.map(_.symbol), vrefss.flatten.map(_.symbol))
+                })
+
+              applyWithForwarding :: specializedApplyDefDef :: acc
+            }
+            .getOrElse(tree :: acc)
+          }
+          case (tree, acc) => tree :: acc
+        }
+
+        val specializedParents = tree.parents.map { t =>
+          symbolMap
+            .get(t.symbol)
+            .map { case (newSym, _) => newSym }
+            .getOrElse(t)
+        }
+
+        cpy.Template(tmpl)(parents = specializedParents, body = body0)
+      }
+      case _ => tree
+    }
+
+  /** Dispatch to specialized `apply`s in user code */
+  override def transformApply(tree: Apply)(implicit ctx: Context, info: TransformerInfo) = {
+    import ast.Trees._
+    tree match {
+      case Apply(select @ Select(id, nme.apply), arg :: Nil) =>
+        val params = List(arg.tpe, tree.tpe)
+        val specializedApply = nme.apply.specializedFor(params, params.map(_.typeSymbol.name), Nil, Nil)
+        val hasOverridenSpecializedApply = id.tpe.decls.iterator.exists { sym =>
+          sym.is(Flags.Override) && (sym.name eq specializedApply)
+        }
+
+        if (hasOverridenSpecializedApply) tpd.Apply(tpd.Select(id, specializedApply), arg :: Nil)
+        else tree
+      case _ => tree
+    }
+  }
+
+  private def specializedName(name: Name, args: List[Type], ret: Type)(implicit ctx: Context): Name = {
+    val typeParams = args :+ ret
+    name.specializedFor(typeParams, typeParams.map(_.typeSymbol.name), Nil, Nil)
+  }
+
+  // Extractors ----------------------------------------------------------------
+  private object ShouldTransformDenot {
+    def unapply(cref: ClassDenotation)(implicit ctx: Context): Option[Seq[SpecializationTarget]] =
```

Comment:
I think it's very simple implementation that doesn't warrant pattern-matching that allocates options.
Additionally, it's used only once. I'd inline it.